File: src/main/java/net/onelitefeather/titan/function/TickleFunction.java
package net.onelitefeather.titan.function;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.minestom.server.entity.Player;
import net.minestom.server.event.Event;
import net.minestom.server.event.EventNode;
import net.minestom.server.event.entity.EntityAttackEvent;
import net.minestom.server.item.Material;
import net.minestom.server.network.packet.server.play.SetCooldownPacket;
import net.minestom.server.tag.Tag;

import java.util.Optional;

public final class TickleFunction {

    private static final Tag<Long> COOLDOWN = Tag.Long("tickle_cooldown");
    private static final String TICKLE_MESSAGE = "<yellow><player> <white>tickled <yellow><target>";

    private final NavigationFunction navigationFunction;

    private void tickleListener(EntityAttackEvent event) {
        if (!(event.getEntity() instanceof Player) || !(event.getTarget() instanceof Player)) return;
        Player player = (Player) event.getEntity();
        Player target = (Player) event.getTarget();
        if (player.getInstance() == null) return;
        if (player.getItemInOffHand().material() == Material.FEATHER || player.getItemInMainHand().material() == Material.FEATHER) {
            if (!player.hasTag(COOLDOWN)) {
                player.setTag(COOLDOWN, System.currentTimeMillis() + 4000);
                SetCooldownPacket packet = new SetCooldownPacket(
                        player.getItemInOffHand().material().id(),
                        (int) ((System.currentTimeMillis() + 4000) / 20)
                );
                player.getPlayerConnection().sendPacket(packet);
                player.getInstance().getPlayers().forEach(worldPlayer -> {
                    Component message = MiniMessage.miniMessage().deserialize(
                            TICKLE_MESSAGE,
                            Placeholder.component("player", Optional.ofNullable(player.getDisplayName()).orElse(player.getName())),
                            Placeholder.component("target", Optional.ofNullable(target.getDisplayName()).orElse(target.getName()))
                    );
                    worldPlayer.sendMessage(message);
                });
            } else if (System.currentTimeMillis() > player.getTag(COOLDOWN)) {
                player.removeTag(COOLDOWN);
            }
        }
    }

    private TickleFunction(EventNode<Event> node, NavigationFunction navigationFunction) {
        this.navigationFunction = navigationFunction;
        node.addListener(EntityAttackEvent.class, this::tickleListener);
    }

    public static TickleFunction instance(EventNode<Event> node, NavigationFunction navigationFunction) {
        return new TickleFunction(node, navigationFunction);
    }

}

File: src/main/kotlin/net/onelitefeather/titan/function/TickleFunction.kt
package net.onelitefeather.titan.function

import com.google.inject.Inject
import com.google.inject.name.Named
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import net.minestom.server.entity.Player
import net.minestom.server.event.Event
import net.minestom.server.event.EventListener
import net.minestom.server.event.EventNode
import net.minestom.server.event.entity.EntityAttackEvent
import net.minestom.server.item.Material
import net.minestom.server.network.packet.server.play.SetCooldownPacket
import net.minestom.server.tag.Tag

class TickleFunction : TitanFunction {

    private val cooldown: Tag<Long> = Tag.Long("tickel_cooldown")
    private val tickleMessage =
        "<yellow><player> <white>tickled <yellow><target>"

    @Named("titanNode")
    @Inject
    lateinit var eventNode: EventNode<Event>

    private fun onEntityAttack(entityAttackEvent: EntityAttackEvent) {
        val player = entityAttackEvent.entity
        val target = entityAttackEvent.target
        if (player is Player && target is Player) {
            player.instance ?: return
            if (player.itemInOffHand.material() == Material.FEATHER || player.itemInMainHand.material() == Material.FEATHER) {
                if (!player.hasTag(cooldown)) {
                    player.setTag(cooldown, System.currentTimeMillis() + 4000)
                    val packet = SetCooldownPacket(
                        player.itemInOffHand.material().id(),
                        ((System.currentTimeMillis() + 4000) / 20).toInt()
                    )
                    player.playerConnection.sendPacket(packet)
                    player.instance?.players?.forEach { worldPlayer ->
                        val message = MiniMessage.miniMessage().deserialize(
                            tickleMessage,
                            Placeholder.component(
                                "player",
                                player.displayName ?: player.name
                            ),
                            Placeholder.component(
                                "target",
                                target.displayName ?: target.name
                            )
                        )
                        worldPlayer.sendMessage(message)
                    }
                } else if (System.currentTimeMillis() > player.getTag(cooldown)) {
                    player.removeTag(cooldown)
                }
            }
        }
    }

    override fun initialize() {
        eventNode.addListener(
            EventListener.of(
                EntityAttackEvent::class.java,
                this::onEntityAttack
            )
        )
    }

    override fun terminate() {
    }

}